[#31] refactor : RestDocs 리팩토링

gombasan opened the pull request:
## 작업 사항

- #31 
로컬에서는 정상적으로 문서화가 되었습니다!!! 목표로 했던 enum 코드 문서화도 진행해봤어요!

[chore : RestDocs 리팩터링](https://github.com/Clover-Habiters/backend/commit/f65771c9127bae68545ab0bcb735d2f315e5d027)
- 기존 문서화를 위한 테스트 코드가 장황 해지는것 같아 코드 리팩토링을 진행하였습니다.
- RestDocsConfig 를 생성하여 RestDocs 의 구성정보를 표현하였습니다. 또한 field 를 추가하여, 속성을 직접 추가할 수 있도록 하였습니다.
- RestDocsSupport 로 RestDocsConfig 를 추가하고, mockMvc 를 직접 주입하여 사용합니다.

[feat : Enum 타입 문서화 추가](https://github.com/Clover-Habiters/backend/commit/6886f594ffd98632c1d707a93d7c93847c2343a2) 
- Enum 코드들을 문서화 하기 위하여 Test 패키지 내에서 사용하는 Controller 를 추가하였습니다.
- Controller 를 호출하여 enum 의 모든 코드들을 반환받아 문서화를 진행합니다.
- 기존 Enum 으로 작성된 코드들의 RestDocsEnum 인터페이스를 상속받아 이름과 설명을 반환하도록 구현하였습니다.
- 해당 이름과 설명으로 문서화를 진행합니다.

[chore : RestDocs 리팩터링 코드 적용](https://github.com/Clover-Habiters/backend/commit/ba3299930ca3807ba176257a953dd43fcab5246a) 
- 기존 Controller 테스트 코드에 RestDocs 내용을 수정하였습니다.

[feat : api adoc 파일 추가 및 수정](https://github.com/Clover-Habiters/backend/commit/bfb724a73ac893ab0ce9823dd657dc311669e2e9) 
- 새로 추가된 api 와 기존에 사용중이던 adoc 파일들을 코드 수정에 맞게 리팩토링하였습니다.
- 기존과 달라진 점은 index 파일로 전부 링크를 걸어서 보도록 변경하였습니다.
- 또한, Enum 코드들은 팝업창으로 띄워 볼 수 있도록 추가하였습니다.

## 이슈 사항

패키지를 일단 한번 정리한다고 했는데.. 제대로 만들었는지가 의문이 드네요! 클래스 이름도 좀 확인이 필요할거같습니다!!
😂😂

gombasan committed:
chore(Test) : ApiDocumentUtils 패키지 변경

- ApiDocumentUtils 클래스를 global.restdocs 로 변경하였습니다.
- 문서화를 위한 유틸 클래스로 해당 패키지가 더 어울리다 생각하여 변경하였습니다.

gombasan committed:
chore : ControllerTest log 설정 추가

- ControllerTest 내에 log4j 를 추가하여 log 추가가 필요할 경우 사용할 수 있도록 수정하였습니다.

gombasan committed:
feat : RestDocs 스니펫 추가

- 문서화를 위한 커스텀 스니펫을 추가하였습니다.
- 기존 request-fields.snippet 안에 제약조건을 추가하였습니다.

gombasan committed:
chore : RestDocs 리팩터링

- 기존 문서화를 위한 테스트 코드가 장황 해지는것 같아 코드 리팩토링을 진행하였습니다.
- RestDocsConfig 를 생성하여 RestDocs 의 구성정보를 표현하였습니다. 또한 field 를 추가하여, 속성을 직접 추가할 수 있도록 하였습니다.
- RestDocsSupport 로 RestDocsConfig 를 추가하고, mockMvc 를 직접 주입하여 사용합니다.
-

gombasan committed:
feat : Enum 타입 문서화 추가

- Enum 코드들을 문서화 하기 위하여 Test 패키지 내에서 사용하는 Controller 를 추가하였습니다.
- Controller 를 호출하여 enum 의 모든 코드들을 반환받아 문서화를 진행합니다.
- 기존 Enum 으로 작성된 코드들의 RestDocsEnum 인터페이스를 상속받아 이름과 설명을 반환하도록 구현하였습니다.
- 해당 이름과 설명으로 문서화를 진행합니다.

gombasan committed:
chore : RestDocs 리팩터링 코드 적용

- 기존 Controller 테스트 코드에 RestDocs 내용을 수정하였습니다.

gombasan committed:
feat : DocumentLinkGenerator 추가

- Enum 코드들의 링크를 만들어주는 클래스를 추가하였습니다.
- popup 이라는 class 를 적용시켜 링크를 클릭했을 경우 popup 창으로 뜨도록 설정하였습니다.

gombasan committed:
feat : docinfo 추가

- 커스텀으로 작성된 자바스크립트를 적용하기위해 docinfo.html 파일을 추가하였습니다.
- 해당 파일은 asciidoctor 가 html 문서를 만들 때, 참조하는 커스텀 파일입니다.
- docinfo 를 참조하여 head 에 스크립트 내용을 생성해주는 역할을 합니다.

gombasan committed:
feat : api adoc 파일 추가 및 수정

- 새로 추가된 api 와 기존에 사용중이던 adoc 파일들을 코드 수정에 맞게 리팩토링하였습니다.
- 기존과 달라진 점은 index 파일로 전부 링크를 걸어서 보도록 변경하였습니다.
- 또한, Enum 코드들은 팝업창으로 띄워 볼 수 있도록 추가하였습니다.

gombasan committed:
config : asciidoctor 수정

- html 파일로 변환 시킬 adoc 파일을 해당 경로에 포함되어있는 모든 adoc 파일을 변환시키도록 변경하였습니다.
- link 를 사용하여 문서화 하기위해 추가하였습니다.

gombasan committed:
delete : 기존 adoc 문서 삭제.

gombasan committed:
chore : 불필요 메서드 삭제

JoosungKwon commented on build.gradle:
```diff
@@ -121,6 +121,9 @@ asciidoctor {
     }
     dependsOn test
     inputs.dir snippetsDir
+    sources {
+        include("**/index.adoc", "**/api/*.adoc", "**/common/*.adoc", "**/enums/*.adoc")
+    }
```
아직 어떤 의미인지 모르겠는데 설명해주실 수 있나요?

gombasan replied:
asciidoctor 가 html 문서를 만들때 사용하는 adoc 파일경로를 직접 정해준거에요!

JoosungKwon replied:
직접 지정해야하나요?

JoosungKwon commented on src/docs/asciidoc/api/Bookmark-API.adoc:
```diff
@@ -0,0 +1,132 @@
+:toclevels: 2
+:sectlinks:
+
+[[Bookmark-API]]
+== 북마크 API
+
+=== 공통
+
+==== 모든 요청은 *Authorization* 헤더가 필요합니다.
+
+include::{snippets}/diary-controller-test/create-diary-test/request-headers.adoc[]
+
+---
```
감사합니다~!!! API문서를 리드미에 연동해도 좋을 것 같아요~

gombasan replied:
머지 시키구 리드미에도 반영하도록 하겠습니다!

JoosungKwon commented on src/docs/asciidoc/api/docinfo.html:
```diff
@@ -0,0 +1,36 @@
+<script>
```
오우~~ 이건 어떻게 하신건가요? 의미는 뭐죠?

gombasan replied:
이게 css 로 popup 이 있는 경우 팝업 창으로 띄우게 해주는 스크립트입니다!
`DocumentLinkGenerator` 클래스에서 `role = popup` 이 부분에서 enum 링크를 popup 이라는 속성을 심어뒀어요!
🫡

JoosungKwon replied:
엄청난 고수...

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/emoji/entity/Emoji.java:
```diff
@@ -63,8 +64,29 @@ public Emoji updateStatus(Type type) {
+		ANGRY("화나요"),
+		SAD("슬퍼요"),
+		SURPRISED("놀라워요"),
+		HEART("하트"),
+		NONE("이모지삭제");
```
해당 두 개의 타입은 제거할 생각입니다!

gombasan replied:
오~ 하트는 뭔가 지금 기획에는 없지만 곧 생길것만 같은 느낌이긴 한데..
코드상으로 남겨두는건 어떨까요?!

JoosungKwon replied:
넵 좋습니닷~

JoosungKwon commented on src/test/java/com/clover/habbittracker/base/ControllerTest.java:
```diff
@@ -37,12 +39,13 @@ public abstract class ControllerTest {
 		savedMember = memberRepository.save(member);
 		accessToken = jwtProvider.createAccessJwt(savedMember.getId());
-		System.out.println("Base Setting Complete");
+		logger.debug("Base Setting Complete");
```
테스트에서는 System.out.println으로도 충분할 것 같았습니다..😅 사실 필요없는 출력이긴 합니다..ㅎ

JoosungKwon commented on src/test/java/com/clover/habbittracker/base/RestDocsSupport.java:
```diff
@@ -0,0 +1,41 @@
+
+import com.clover.habbittracker.global.restdocs.config.RestDocsConfig;
+
+@Disabled
```
왜 Disabled 인가요?

gombasan replied:
아 혹시 해당 RestDocsSupport 가 테스트로 실행되는게 아닌지 생각이 들어서.. 잠깐 넣어뒀떤건데 남아있네요!
제거하도록 하겠씁니다!

JoosungKwon commented on src/test/java/com/clover/habbittracker/domain/bookmark/api/BookmarkControllerTest.java:
```diff
@@ -94,17 +58,18 @@ void createBookmark() throws Exception {
-					fieldWithPath("description").type(STRING).description("북마크 설명")
+					fieldWithPath("title").type(STRING)
+						.description("북마크 제목")
+						.attributes(field("constraints", "30자 이내")),
```
굳~~😆

gombasan replied:
🙏

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/restdocs/common/CommonDocController.java:
```diff
@@ -0,0 +1,41 @@
+import com.clover.habbittracker.global.restdocs.enums.EnumResponse;
+
+@RestController
+@RequestMapping("/test")
```
테스트보다` /docs` 가 나을 것 같아요!

gombasan replied:
오... 그런 더 좋은 이름이 있었군요... 감사합니다 ~~!!

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/restdocs/config/RestDocsConfig.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation;
+import org.springframework.restdocs.mockmvc.RestDocumentationResultHandler;
+
+@TestConfiguration
+public class RestDocsConfig {
+
+	@Bean
+	public RestDocumentationResultHandler write() {
+		return MockMvcRestDocumentation.document(
+			"{class-name}/{method-name}",
+			getDocumentRequest(),
+			getDocumentResponse()
+		);
+	}
+
+	public static Attribute field(
+		String key,
+		String value) {
+		return new Attribute(key, value);
+	}
+}
```
LGTM 👍

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/restdocs/enums/EnumDocs.java:
```diff
@@ -0,0 +1,71 @@
+	private Map<String, String> emojiType;
+	private Map<String, String> category;
+
+	private EnumDocs() {
+		this(null, null, null);
+	}
```
?

```
this.emojiDomain = new HashMap<>();
this.category = new HashMap<>();
this.emojiType = new HashMap<>();
```

이렇게 두면 안되나요?

gombasan replied:
맞네요...ㅋㅋㅋㅋㅋㅋㅋ 후딱 수정하도록 하겠습니다!!

JoosungKwon reviewed (approved):
@gombasan 태희 형님 고생많으셨습니다!!! 👍

딱봐도 쉽지 않았을텐데 고생하셨네요~ 다음번 미팅에서 부연 설명 좀 부탁드리겠습니다~ 머지해도 될것 같아요~!! 그리고 RestDocsSupport를 상속받는 구조가 좀 애매하다고 생각되는데 다른 방식은 어려울까요?

gombasan committed:
chore : 피드백 내용 반영

- CommonDocController url 주소를 수정하였습니다.
- EnumDocs 의 잘못된 기본 생성자를 수정하였습니다.
- RestDocsSupport @Disabled 어노테이션을 삭제하였습니다.

gombasan commented:
[chore : 피드백 내용 반영](https://github.com/Clover-Habiters/backend/pull/32/commits/94f0b986712a728466e8043ccd7edf1364860399)

위에 수정사항 반영 하였습니다~~!!
머지 하도록 할께요!
